rfc/rfc-77/rfc-77.md

```diff
@@ -0,0 +1,323 @@
+```
+
+Merged Records (sorted):
+``` 
+austin -> 9809a8b1-2d15-4d3d-8ec9-efc48c536a01
+chennai -> e3cf430c-889d-4015-bc98-59bdce1e530c
+chennai -> c8abbe79-8d89-47ea-b4ce-4d224bae5bfa
+los-angeles -> 9909a8b1-2d15-4d3d-8ec9-efc48c536a01
+sfo -> 334e26e9-8355-45cc-97c6-c31daf0df330
+```
+
+An important detail to be noted here is that grouping for merging depends on the `secondary-key, record-key` tuple (`city, uuid` 
+in this example). This is radically different from any other existing partitions in Hudi (MDT or base tables). Hence, 
+support needs to be added to merge records from MDT partition hosting SI (with generalized approach that can be used for
+any future non-unique record key based dataset).
+
+The data structure built for holding the merged log records should enable to search similar keys efficiently with 
+following consideration:
+1. Should allow spilling to disk (to prevent OOM). A pure in-memory data structure will not be production ready, but 
+can be used only for POC
+2. Should allow for searching (and retrieving) based on two keys as efficiently as possible. First search will be based 
+on `secondary-key` and second search will be based on `record-key`. Hence, a single level Map used by 
+`HoodieMergedLogRecordScanner` will not work.
+3. Should allow for efficient removal of records (deleted record need to be removed). Hence, any data structure that 
+uses a flat array will not be efficient. 
+4. Should allow for efficient insertion of records (for inserting merged record and for buffering fresh records). 
+
+The [initial POC](https://github.com/apache/hudi/pull/10625) makes use of an in-memory nested maps - with the first level keyed by `secondary-key`  
+and the second level keyed by`record-key`. However, the final design should allow spilling to disk.
+
+Considering the above requirements, the proposal is to extend `HoodieKey` to support secondary keys. We will make `HoodieKey` generic and create a subclass `HoodieSecondaryKey` where the key can be a generic type.
+```java
+public class HoodieKey<T extends Serializable> implements Serializable {
+
+  private T recordKey;
+  private String partitionPath;
+
+  // Required for serializer
+  public HoodieKey() {}
+
+  public HoodieKey(T recordKey, String partitionPath) {
+    this.recordKey = recordKey;
+    this.partitionPath = partitionPath;
+  }
+
+  // Getters and setters
+  public T getRecordKey() {
+    return recordKey;
+  }
+
+  public void setRecordKey(T recordKey) {
+    this.recordKey = recordKey;
+  }
+
+  public String getPartitionPath() {
+    return partitionPath;
+  }
+
+  public void setPartitionPath(String partitionPath) {
+    this.partitionPath = partitionPath;
+  }
+}
+```
+
+Since `HoodieKey` is now generic, `HoodieSecondaryKey` can specify any type for its key by setting the type parameter
+when it extends `HoodieKey`.
+
+```java
+public class HoodieSecondaryKey<K extends Serializable> extends HoodieKey<K> {
```

Comment:
lets make a new class hierarchy. `MergeKey` (interface) and then implemented by `SimpleMergeKey` and `CompositeMergeKey` etc.

Reply:
Agreed. Updated the RFC with the details on the new class hierarchy.